src/puppetlabs/puppetdb/utils.clj

```diff
@@ -134,3 +134,36 @@
     (let [[certname-facts more-rows] (split-with (split-pred rows) rows)]
       (cons (collapse-fn certname-facts)
             (lazy-seq (collapse-seq split-pred collapse-fn more-rows))))))
+
+(def base-url-schema
+  {:protocol s/Str
+   :host s/Str
+   :port s/Int
+   (s/optional-key :prefix) s/Str
+   (s/optional-key :version) (s/both
+                              s/Keyword
+                              (s/pred #(and (keyword? %)
+                                            (re-matches #"v\d+" (name %)))
+                                      'valid-version?))})
+
+(pls/defn-validated base-url->str :- s/Str
+  "Converts the `base-url' map to an ASCII URL.  May throw
+   MalformedURLException or URISyntaxException."
+  [base-url :- base-url-schema]
+  (-> (URL. (:protocol base-url) (:host base-url) (:port base-url)
+            (str (when-let [p (:prefix base-url)]
+                   (str "/" p))
+                 "/" (name (or (:version base-url) :v4))))
+    .toURI .toASCIIString))
```

Comment:
It's just a hack for the error checking side-effect -- I may have overlooked it, but I didn't see a way to get URL/URI to do that otherwise.